File: 2023/UICPC Regional/Round #9/solutions/traffic/submissions/accepted/traffic_zf.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int aa[100000], bb[100000];

int main() {
    int x1, x2, n1, n2;
    int *a = aa, *b = bb;
    cin >> x1 >> x2;

    cin >> n1;
    for (int i = 0; i < n1; ++i) cin >> a[i];

    cin >> n2;
    for (int i = 0; i < n2; ++i) cin >> b[i];

    if (x1 > x2) {
        swap(x1, x2);
        swap(n1, n2);
        swap(a, b);
    }

    int i = 0, j = 0, d1 = 0, d2 = 0, t = 0, dt;

    while (i < n1 || j < n2) {
        bool first = (j == n2 || (i < n1 && a[i] < b[j]));

        dt = (first ? a[i]-t : b[j]-t);
        int nx1 = x1+dt*d1, nx2 = x2+dt*d2;

        if (nx1 + 5 > nx2) {
            cout << "bumper tap at time " << x2-x1-4+t << endl;
            return 0;
        }

        x1 = nx1;
        x2 = nx2;

        if (first) { d1 = 1-d1; ++i; }
        else       { d2 = 1-d2; ++j; }

        t += dt;
    }

    if (d1 == 1 && d2 == 0) cout << "bumper tap at time " << x2-x1-4+t << endl;
    else cout << "safe and sound" << endl;
    return 0;
}

File: 2023/UICPC Regional/Round #9/solutions/traffic/submissions/accepted/traffic_zf.py
#!/usr/bin/python3

def solve(x1, x2, t1, t2):
    if x1 > x2:
        x1, x2 = x2, x1
        t1, t2 = t2, t1

    t1.append(10**9)
    t2.append(10**9)

    i1, i2, ct, d1, d2, crash = 0, 0, 0, 0, 0, -1

    while i1 < len(t1) and i2 < len(t2):
        delta = min(t1[i1], t2[i2]) - ct
        xx1 = x1 + delta*d1
        xx2 = x2 + delta*d2

        if xx1 + 5 > xx2: return ct + x2 - x1 - 4

        x1, x2 = xx1, xx2
        ct += delta

        if t1[i1] < t2[i2]:
            d1 = 1-d1
            i1 += 1
        else:
            d2 = 1-d2
            i2 += 1

    return -1


x1,x2 = map(int, input().split())

t1 = list(map(int, input().split()))
t2 = list(map(int, input().split()))

crash = solve(x1, x2, t1[1:], t2[1:])
if crash == -1: print("safe and sound")
else: print("bumper tap at time " + str(crash))
